LDH/Day15/계산기.js

```diff
@@ -0,0 +1,120 @@
+const exprForm = document.getElementById("calc-form")
+    .addEventListener("submit", function (e) {
+        e.preventDefault(); // 폼 제출 막음
+
+        const exprInput = document.getElementById("expression");
+        const userText = exprInput.value.trim();
+
+        if (checkdateInput(userText)) {
+            main(userText);
+        }
+    });
+
+function checkdateInput(userText) {
+    if (userText === "exit") {
+        alert("종료");
+        return false;
+    }
+    if (userText === "") {
+        alert("값이 없습니다.");
+        return false;
+    }
+    const Values = /^[\d+\-*/() ]+$/;
+    if (!Values.test(userText)) {
+        alert("잘못된 수식입니다.");
+        return false;
+    }
+    return true;
+}
+
+function main(userText) {
+    const tokens = parseExpr(userText);
+
+    const convertedTokens = tokenize(tokens);
+
+    const postfix = calcPriority(convertedTokens);
+
+    const [result] = evaluatePostfix(postfix);
+    
+    showResult(result);
+}
+
+// 문자열 파싱 => 숫자/연산자 배열
+function parseExpr(tokens) {
+    const tokensResult = tokens.match(/\d+|[+\-*/()]/g);
+    return tokensResult;
+}
+
+// 숫자 문자열은 숫자로 변환
+function tokenize(tokens) {
+    return tokens.map(a => isNaN(a) ? a : Number(a));
+}
+
+// 후위 표기식 변환
+function calcPriority(tokens) {
+    const output = [];
+    const stack = [];
+    const priority = {
+        '+': 1,
+        '-': 1,
+        '*': 2,
+        '/': 2,
+    };
+
+    for (let token of tokens) {
+        if (typeof token === 'number') {
+            output.push(token);
+        } else {
+            while (
+                stack.length > 0 &&
+                priority[stack[stack.length - 1]] >= priority[token]
+            ) {
+                output.push(stack.pop());
+            }
+            stack.push(token);
+        }
+    }
+
+    while (stack.length > 0) {
+        output.push(stack.pop());
+    }
+
+    return output;
+}
+
+// 후위 표기식 계산
+function evaluatePostfix(postfix) {
+    const stack = [];
+
+    for (let token of postfix) {
+        if (typeof token === 'number') {
+            stack.push(token)
+        } else {
+            let left = stack.pop();
+            let right = stack.pop();
+            stack.push(calculate(left, right, token));
```

Comment:
스택 구조상 right를 먼저 pop해야 연산 순서가 맞습니다.
2 - 5 → calculate(2, 5, '-')가 되어야 하는데 현재는 calculate(5, 2, '-')로 잘못 계산되고 있습니다.
```suggestion
            let right = stack.pop();
            let left = stack.pop();
            stack.push(calculate(left, right, token));
```

Reply:
처음에 right - left나 right / left로 계산했더니 값이 이상하게 나와서 right - left랑 right / left해서 계산을 했었는데 그게 스택의 후입선출 구조 때문에 연산 순서가 바뀌어서 그런 거였네요. 알려주셔서 감사합니다!